h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstSpearman.java

```diff
@@ -0,0 +1,125 @@
+package water.rapids.ast.prims.advmath;
+
+import water.DKV;
+import water.MRTask;
+import water.Value;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+import water.fvec.Vec;
+import water.rapids.Env;
+import water.rapids.Merge;
+import water.rapids.Val;
+import water.rapids.ast.AstPrimitive;
+import water.rapids.ast.AstRoot;
+import water.rapids.ast.prims.mungers.AstRankWithinGroupBy;
+import water.rapids.vals.ValNum;
+
+public class AstSpearman extends AstPrimitive<AstSpearman> {
+  @Override
+  public int nargs() {
+    return 1 + 3; // Frame ID and ID of two numerical vectors to calculate SCC on.
+  }
+
+  @Override
+  public String[] args() {
+    return new String[0];
+  }
+
+  @Override
+  public Val apply(Env env, Env.StackHelp stk, AstRoot[] asts) {
+    final Value value = DKV.get(asts[1].str());
+    if (!value.isFrame()) {
+      throw new IllegalArgumentException(String.format("The given key '%s' is not a frame.", asts[1]));
+    }
+
+    final Frame originalUnsortedFrame = value.get(Frame.class);
+    final int vecIdX = originalUnsortedFrame.find(asts[2].str());
+    final int vecIdY = originalUnsortedFrame.find(asts[3].str());
+
+    final Frame unsortedFrameWithoutNAs = new Merge.RemoveNAsTask(vecIdX, vecIdY)
+            .doAll(originalUnsortedFrame.types(), originalUnsortedFrame)
+            .outputFrame(originalUnsortedFrame.names(), originalUnsortedFrame.domains());
+
+    // Sort by X
+    final AstRankWithinGroupBy.SortnGrouby sortTaskX = new AstRankWithinGroupBy.SortnGrouby(unsortedFrameWithoutNAs, new int[]{},
+            new int[]{vecIdX}, new int[]{1}, "rankX");
+    sortTaskX.doAll(sortTaskX._groupedSortedOut);
+
+    AstRankWithinGroupBy.RankGroups rankTaskX = new AstRankWithinGroupBy.RankGroups(sortTaskX._groupedSortedOut, sortTaskX._groupbyCols,
+            sortTaskX._sortCols, sortTaskX._chunkFirstG, sortTaskX._chunkLastG, sortTaskX._newRankCol)
+            .doAll(sortTaskX._groupedSortedOut);
+
+
+    // Sort by Y
+    final AstRankWithinGroupBy.SortnGrouby sortTaskY = new AstRankWithinGroupBy.SortnGrouby(rankTaskX._finalResult, new int[]{},
+            new int[]{vecIdY}, new int[]{1}, "rankY");
+    sortTaskY.doAll(sortTaskY._groupedSortedOut);
+
+    AstRankWithinGroupBy.RankGroups rankTaskY = new AstRankWithinGroupBy.RankGroups(sortTaskY._groupedSortedOut, sortTaskY._groupbyCols,
+            sortTaskY._sortCols, sortTaskY._chunkFirstG, sortTaskY._chunkLastG, sortTaskY._newRankCol)
+            .doAll(sortTaskY._groupedSortedOut);
+
+    final Vec rankX = rankTaskY._finalResult.vec("rankX");
```

Comment:
Currently, the ranks have two problems:

- Duplicated values are given different rank (they're sorted one after another correctly, but their rank is incremented)
- There are "gaps" and "holes" between the ranks. E.g. `1,2,5,6`.

Reply:
When I'm thinking about it, the way we need duplicated values treated requires cross-chunk checking. For the computation form. There is alternative way, just by adjusting the equation, we might only need sorted values, not ranked.